skfem/assembly/form/coo_data.py

```diff
@@ -0,0 +1,54 @@
+from dataclasses import dataclass, replace
+from typing import Tuple
+
+import numpy as np
+from numpy import ndarray
+from scipy.sparse import coo_matrix, csr_matrix
+
+
+@dataclass
+class COOData:
+
+    data: ndarray
+    rows: ndarray
+    cols: ndarray
+    shape: Tuple[int, int]
+
+    @staticmethod
+    def _assemble_scipy_csr(data: ndarray,
+                            rows: ndarray,
+                            cols: ndarray,
+                            shape: Tuple[int, int]) -> csr_matrix:
+
+        K = coo_matrix((data, (rows, cols)), shape=shape)
+        K.eliminate_zeros()
+        return K.tocsr()
+
+    def tocsr(self) -> csr_matrix:
+
+        return self._assemble_scipy_csr(
+            self.data,
+            self.rows,
+            self.cols,
+            self.shape,
+        )
+
+    def enforce(self, D):
+        """Enforce an essential BC by setting rows and diagonals to 0 and 1.
+
+        Parameters
+        ----------
+        D
+            A list of (Dirichlet) degrees-of-freedom to enforce.
+
+        """
+        rows_mapping = np.ones(self.shape[0])
```

Comment:
Should this method have an option for handling mass matrices?  Or &lsquo;pencils&rsquo; consisting of a stiffness matrix and a mass matrix? As in Mx' + Kx = 0, as in 

https://github.com/kinnala/scikit-fem/blob/a88739744505ace72f59021400a548f3e16e0ded/docs/examples/ex29.py#L122

Or more general cases like the second-order Mx'' + Cx' + Kx = f, e.g. for linear elastodynamics.  I think that M or M & C get zeros on the diagonal rather than ones, for constant Dirichlet data.

I've been meaning to put together a simple example to investigate #591 in generalized eigenproblems, e.g. a vibrating string fixed at each end.  (Not ex16 as that doesn't need boundary conditions because the termini are singular points of Legendre's equation.)

Reply:
So basically skipping the part where diagonal is set to zero?

Reply:
Check out the latest commit.